src/main/java/org/folio/circulation/resources/ItemsInTransitResource.java

```diff
@@ -40,21 +63,170 @@ private void getMany(RoutingContext routingContext) {
     final WebContext context = new WebContext(routingContext);
     final Clients clients = Clients.create(context, client);
 
+    final CollectionResourceClient loansStorageClient = clients.loansStorage();
+    final RequestRepository requestRepository = RequestRepository.using(clients);
     final ItemRepository itemRepository = new ItemRepository(clients, true, true, true);
+    final ServicePointRepository servicePointRepository = new ServicePointRepository(clients);
+    final ReportRepository reportRepository = new ReportRepository(clients);
 
-    CompletableFuture<Result<Collection<Item>>> itemsInTransit = itemRepository
-      .findByQuery(exactMatch("status.name", IN_TRANSIT.getValue()));
-
-    itemsInTransit
+    reportRepository.getAllItemsByField("status.name", IN_TRANSIT.getValue())
+      .thenComposeAsync(r -> r.after(itemsReportFetcher ->
+        fetchInTransitReportEntry(itemsReportFetcher, itemRepository, servicePointRepository)))
+      .thenComposeAsync(r -> r.after(inTransitReportEntries ->
+        fetchLoans(loansStorageClient, servicePointRepository, inTransitReportEntries)))
+      .thenComposeAsync(r -> findRequestsByItemsIds(requestRepository, r.value()))
       .thenApply(this::mapResultToJson)
       .thenApply(OkJsonResponseResult::from)
       .thenAccept(r -> r.writeTo(routingContext.response()));
   }
 
-  private Result<JsonObject> mapResultToJson(Result<Collection<Item>> items) {
-    return items.map(resultList -> resultList
-      .stream()
-      .map(item -> new ItemReportRepresentation().createItemSummary(item))
+  public CompletableFuture<Result<List<InTransitReportEntry>>> fetchInTransitReportEntry(ItemsReportFetcher itemsReportFetcher,
+                                                                                         ItemRepository itemRepository,
+                                                                                         ServicePointRepository servicePointRepository) {
+    List<InTransitReportEntry> inTransitReportEntries = itemsReportFetcher.getResultListOfItems().stream()
+      .flatMap(records -> records.value().getRecords()
+        .stream()).map(item -> fetchRelatedRecords(itemRepository, servicePointRepository, item))
+      .map(CompletableFuture::join)
```

Comment:
From my experience calling `join` method can cause deadlocks (it will block the event loop thread). I'd use `thenAccept/Apply` and callbacks here instead.

But, maybe I'm missing something.

But if you don't observe any timeouts/deadlocks in unit tests, then I think it will be fine using the `join` method.

Reply:
I agree with you @bohdan-suprun. Even without timeouts/deadlocks in the unit tests, we should not be doing a join here. If we need to block, we should defer to Vert.x's blocking mechanisms and, even then, I don't think we need to block to make this code work.

Reply:
Is there a way of doing this with `allOf` and flattening the list of `CompletableFutures` after it completes?

Reply:
Changed